drivers/spi/spi_mchp_mec5_qspi.c

```diff
@@ -0,0 +1,539 @@
+		xlen = spi_context_max_continuous_chunk(ctx);
+		data->chunk_size = xlen;
+
+		uint8_t const *txb = ctx->tx_buf;
+		uint8_t *rxb = ctx->rx_buf;
+
+		if (txb) {
+			qflags |= MEC5_QSPI_ULDMA_FLAG_INCR_TX;
+		} else {
+			txb = &devcfg->ovrc;
+		}
+
+		if (rxb) {
+			qflags |= MEC5_QSPI_ULDMA_FLAG_INCR_RX;
+		} else {
+			rxb = (uint8_t *)&data->rxdb;
+		}
+
+		if ((data->total_tx_size <= xlen) && (data->total_rx_size <= xlen)) {
+			qflags |= MEC5_QSPI_ULDMA_FLAG_CLOSE;
+		}
+
+		data->xfr_flags = MEC5_QSPI_XFR_FLAG_LDMA;
+		mec_hal_qspi_uldma_fd2(regs, (const uint8_t *)txb, rxb, xlen, qflags);
+	} else {
+		spi_context_complete(&data->ctx, dev, 0);
+	}
+}
+
+static void mec5_qspi_isr(const struct device *dev)
+{
+	struct mec5_qspi_data *data = dev->data;
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	uint32_t hwsts = 0u;
+	int status = 0;
+
+	hwsts = mec_hal_qspi_hw_status(regs);
+	data->qstatus = hwsts;
+	status = mec_hal_qspi_done(regs);
+
+	mec_hal_qspi_intr_ctrl(regs, 0);
+	mec_hal_qspi_hw_status_clr(regs, hwsts);
+	mec_hal_qspi_girq_clr(regs);
+
+	if ((data->xfr_flags & MEC5_QSPI_XFR_FLAG_START) && (hwsts & MEC_QSPI_IEN_TXB_EMPTY)) {
+		/* empty TX FIFO mec_hal_qspi_done returns BUSY. Ignore if starting transfer */
+		data->xfr_flags &= (uint32_t)~MEC5_QSPI_XFR_FLAG_START;
+	} else if (status != MEC_RET_OK) {
+		status = -EIO;
+		spi_context_complete(&data->ctx, dev, status);
+	}
+
+	mec5_qspi_ctx_next(dev);
+}
+
+/*
+ * Called for each QSPI controller by the kernel during driver load phase
+ * specified in the device initialization structure below.
+ * Initialize QSPI controller.
+ * Initialize SPI context.
+ * QSPI will be fully configured and enabled when the transceive API
+ * is called.
+ */
+static int mec5_qspi_init(const struct device *dev)
+{
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	struct mec5_qspi_data *data = dev->data;
+	enum mec_qspi_cs cs = MEC_QSPI_CS_0;
+	enum mec_qspi_io iom = MEC_QSPI_IO_FULL_DUPLEX;
+	enum mec_qspi_signal_mode spi_mode = MEC_SPI_SIGNAL_MODE_0;
+	int ret = 0;
+
+	data->cs = 0;
+
+	ret = mec_hal_qspi_init(regs, (uint32_t)devcfg->clock_freq, spi_mode, iom, cs);
+	if (ret != MEC_RET_OK) {
+		LOG_ERR("QSPI init error (%d)", ret);
+		return -EINVAL;
+	}
+
+	data->freq = devcfg->clock_freq;
+	data->operation = SPI_WORD_SET(8) | SPI_LINES_SINGLE;
+	mec_hal_qspi_byte_time_ns(regs, &data->byte_time_ns);
+
+	ret = pinctrl_apply_state(devcfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		LOG_ERR("QSPI pinctrl setup failed (%d)", ret);
+	}
+
+	ret = spi_context_cs_configure_all(&data->ctx);
+	if (ret) {
+		LOG_ERR("QSPI cs config failed (%d)", ret);
+		return ret;
+	}
+
+	if (devcfg->irq_config_func) {
+		devcfg->irq_config_func();
+	}
+
+	spi_context_unlock_unconditionally(&data->ctx);
+
+	return ret;
+}
+
+static const struct spi_driver_api mec5_qspi_driver_api = {
```

Comment:
Use the device API wrapper macro.

```suggestion
static DEVICE_API(spi, mec5_qspi_driver_api) = {
```

Reply:
@pdgendt Hi, thanks for the review.  Change your recommended implemented and update pushed to PR.